### PR TITLE
server: retry store id allocation on retryable failure

### LINKED ISSUE #12955: teamcity: Fatal WriteTooOldError when allocating store IDs in test
The following tests appear to have failed:

[#120518](https://teamcity.cockroachdb.com/viewLog.html?buildId=120518):

```
--- FAIL: acceptance/TestBuildInfo (48.331s)
	soon.go:43: acceptance/util.go:422, condition failed to evaluate within 45s: expected 3 replicas, only found 2
	localcluster.go:701: unexpected extra event &{1 die} (after [])
	test_log_scope.go:81: test log files left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestBuildInfo127467030

```
Please assign, take a look and update the issue accordingly.

### PR DESCRIPTION
**Pull request by andreimatei**
fixes #12955

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13107)

### CODE REVIEW

File: pkg/server/node.go [hunk 1]
```diff
@@ -150,7 +150,18 @@ func allocateNodeID(ctx context.Context, db *client.DB) (roachpb.NodeID, error)
...
 ) (roachpb.StoreID, error) {
-	r, err := db.Inc(ctx, keys.StoreIDGenerator, inc)
+	var err error
+	var r client.KeyValue
+	// Perform a retryable non-transactional increment.
+	for retry := true; retry; {
+		r, err = db.Inc(ctx, keys.StoreIDGenerator, inc)
+		switch err.(type) {
+		default:
+			retry = false
```
Comment (tamird) [inline pkg/server/node.go]: isn't this just `break`?

Comment (andreimatei) [inline pkg/server/node.go]: break breaks out of the `switch`...

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/server/node.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbD2Q59M3kBygheGmLE:-KbD2Q5AUWSvT7wMVM1x:b-rbxohv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3892d305c65734658f178dc6a0382ab26dea9dae/pkg/server/node.go#L162)):*
> ```Go
> 			retry = false
> 		case *roachpb.WriteTooOldError, *roachpb.AmbiguousResultError:
> 			retry = true
> ```

`retry` is already true here.

---

*[pkg/server/node.go, line 547 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbD2AmsayVNfXztCZYF:-KbD2AmsayVNfXztCZYG:b-cn45y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3892d305c65734658f178dc6a0382ab26dea9dae/pkg/server/node.go#L547)):*
> ```Go
> 	firstID, err := allocateStoreIDs(ctx, n.Descriptor.NodeID, inc, n.storeCfg.DB)
> 	if err != nil {
> 		log.Fatalf(ctx, "error allocating store ids: %+v", err)
> ```

What is `%+v` doing here? I think `%v` would be more appropriate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-:-KbD2TJ14r9uDLXn538p:bnfp4nl)*

File: pkg/server/node.go (hunk 1 above)
Comment (tamird) [inline pkg/server/node.go]:
```
for {
  r, err = db.Inc(ctx, keys.StoreIDGenerator, inc)
  switch err.(type) {
  case *roachpb.WriteTooOldError, *roachpb.AmbiguousResultError:
    continue
  }
  break
}
```

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/server/node.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbD1BoRQenSvFCu83tD-r1-160:-KbD4F04nmKRsjtDCI_e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3892d305c65734658f178dc6a0382ab26dea9dae/pkg/server/node.go#L160)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

```
for {
  r, err = db.Inc(ctx, keys.StoreIDGenerator, inc)
  switch err.(type) {
  case *roachpb.WriteTooOldError, *roachpb.AmbiguousResultError:
    continue
  }
  break
}
```
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbD2Q59M3kBygheGmLE:-KbD4GaBeUklFG2LqWUq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3892d305c65734658f178dc6a0382ab26dea9dae/pkg/server/node.go#L162)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`retry` is already true here.
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 547 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbD2AmsayVNfXztCZYF:-KbD4HLKKXJ_Xrx-p36w:bq5vu2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3892d305c65734658f178dc6a0382ab26dea9dae/pkg/server/node.go#L547)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What is `%+v` doing here? I think `%v` would be more appropriate.
</blockquote></details>

`%+v` will print the stack of the error

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-:-KbD5cWZl4k51HFZL3Zd:bnfp4nl)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/server/node.go, line 547 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbD2AmsayVNfXztCZYF:-KbD6SUkft2rDtVjaSTD:b-j284ht) ([raw file](https://github.com/cockroachdb/cockroach/blob/3892d305c65734658f178dc6a0382ab26dea9dae/pkg/server/node.go#L547)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`%+v` will print the stack of the error
</blockquote></details>

It sure does. Didn't realize you wanted that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107)*

**Comment by a-robinson** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/node.go, line 139 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbFubf4OzV4ibLDeH6T:-KbFubf4OzV4ibLDeH6U:b-ejh4h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/962cfcf66165b49411761561cfed0e6d78e5d270/pkg/server/node.go#L139)):*
> ```Go
> // allocateNodeID increments the node id generator key to allocate
> // a new, unique node id.
> func allocateNodeID(ctx context.Context, db *client.DB) (roachpb.NodeID, error) {
> ```

We're going to need the same logic here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107)*

**Comment by bdarnell** [comment]
LGTM, but I wonder if we should just export and reuse `storage.idAllocator` (possibly moving it to the `client` package). It does more than we need here since it allocates IDs in blocks, but it would centralize the retry logic in one place.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/server/node.go, line 156 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbG1DD0FRoOoSlR46uR:-KbG1DD0FRoOoSlR46uS:b4gyzw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/962cfcf66165b49411761561cfed0e6d78e5d270/pkg/server/node.go#L156)):*
> ```Go
> 	var r client.KeyValue
> 	// Perform a retryable non-transactional increment.
> 	for {
> ```

All retry loops should have a backoff. Use util/retry here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-:-KbG1Kumi2JWgs5JxcHd:bh95039)*

**Comment by andreimatei** [comment]
I think `storage.idAllocator` is too heavy-weight; I didn't use it.
A related question is whether we should retry on all errors or only expected retryable ones. `idAllocator` retries on all. I chose to only retry on WriteTooOld and Ambiguous in the case of node id generation, since this is something that only happens at startup and crashing at startup is not the worst thing.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/server/node.go, line 139 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbFubf4OzV4ibLDeH6T:-KbH4HrnMm3-_U8vgm_a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/962cfcf66165b49411761561cfed0e6d78e5d270/pkg/server/node.go#L139)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

We're going to need the same logic here.
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 156 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbG1DD0FRoOoSlR46uR:-KbH4IoSWE0Jpe5vA7ir:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/962cfcf66165b49411761561cfed0e6d78e5d270/pkg/server/node.go#L156)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

All retry loops should have a backoff. Use util/retry here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-:-KbH4N3F_wxGiBzZ4HSb:bz3r6ox)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/server/node.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbH67FWDWbFnUYDdH_S:-KbH67FWDWbFnUYDdH_T:b-spfptk) ([raw file](https://github.com/cockroachdb/cockroach/blob/951bd24823f732698681d3a0d4603a3e19f85b21/pkg/server/node.go#L142)):*
> ```Go
> 	val, err := incVal(ctx, db, keys.NodeIDGenerator, 1)
> 	if err != nil {
> 		return 0, errors.Wrapf(err, "unable to allocate node ID")
> ```

`Wrap`, not `Wrapf`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107)*

**Comment by a-robinson** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-:-KbH7I4BAxbbdALMDyIe:bnfp4nl)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/server/node.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107#-KbH67FWDWbFnUYDdH_S:-KbHBDcooSvEQ-_hayKL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/951bd24823f732698681d3a0d4603a3e19f85b21/pkg/server/node.go#L142)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`Wrap`, not `Wrapf`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13107)*

**Pull request merged**